main-settings/src/main/scala/sbt/Scope.scala

```diff
@@ -266,36 +266,65 @@ object Scope {
   )(rawScope: Scope): Seq[Scope] = {
     val scope = Scope.replaceThis(GlobalScope)(rawScope)
 
-    def nonProjectScopes(resolvedProj: ResolvedReference)(px: ScopeAxis[ResolvedReference]) = {
-      val p = px.toOption getOrElse resolvedProj
-      val configProj = p match {
-        case pr: ProjectRef => pr; case br: BuildRef => ProjectRef(br.build, rootProject(br.build))
-      }
-      val cLin = scope.config match {
-        case Select(conf) => index.config(configProj, conf); case _ => withZeroAxis(scope.config)
-      }
+    // This is a hot method that gets called many times
+    def exandDelegateScopes(resolvedProj: ResolvedReference)(
+        pLin: Seq[ScopeAxis[ResolvedReference]]): Vector[Scope] = {
       val tLin = scope.task match {
-        case t @ Select(_) => linearize(t)(taskInherit); case _ => withZeroAxis(scope.task)
+        case t @ Select(_) => linearize(t)(taskInherit)
+        case _             => withZeroAxis(scope.task)
+      }
+      // val eLin = withZeroAxis(scope.extra)
+      // The following while loops handroll the nested for-expression + flatMap
+      // projAxes flatMap nonProjectScopes(resolvedProj)
+      //   ...
+      //   for (c <- cLin; t <- tLin; e <- eLin) yield Scope(px, c, t, e)
+      val res = Vector.newBuilder[Scope]
```

Comment:
I'm not sure `Vector` is the best choice here. `ListBuffer` would be the same as used before, it might make sense to use that first, and the separately try a different `SeqBuilder`. When I was looking at speeding up this code, I tried a pre-sized `ArrayBuffer`, I think.

Reply:
But a big 👍 for getting rid of the nested `flatMap`-s, every level of nesting doubles the amount of garbage created.

Reply:
The ideal structure would actually be a hand-rolled `Iterator[Scope]`, but is a huge pain to refactor into that form.

Reply:
I went for `Vector.newBuilder` based on https://stackoverflow.com/a/14059155/3827. @jrudolph went for the presized `val res = new ArrayBuffer[Scope](pLin.size * 2 * tLin.size * eLin.size)`, but I didn't see observable difference, so I figured that no longer became the bottleneck.

Reply:
> The ideal structure would actually be a hand-rolled `Iterator[Scope]`, but is a huge pain to refactor into that form.

That might be a good idea for hackathon to add an iterator version.

Reply:
> The ideal structure would actually be a hand-rolled Iterator[Scope], but is a huge pain to refactor into that form.

:+1: